File: sdk/java/src/main/java/com/experimentationplatform/sdk/FeatureFlagEvaluator.java
package com.experimentationplatform.sdk;

import com.experimentationplatform.sdk.model.FeatureFlag;
import com.experimentationplatform.sdk.model.User;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.List;

/**
 * Evaluates feature flags locally using consistent hashing.
 *
 * <h2>Algorithm</h2>
 * <p>The hash algorithm is byte-for-byte compatible with the Python Lambda
 * implementation in {@code backend/lambda/shared/consistent_hash.py}:
 * <ol>
 *   <li>Compute MD5 of {@code "{userId}:{salt}"} encoded as UTF-8.</li>
 *   <li>Interpret the first 4 bytes as a <strong>little-endian</strong> unsigned 32-bit integer.</li>
 *   <li>Divide by {@code 0x100000000L} (2^32) to normalize to [0.0, 1.0).</li>
 * </ol>
 *
 * <p>The same input always produces the same normalized value, ensuring deterministic
 * and consistent assignment across all SDK implementations (Java, Python, JS).
 *
 * <h2>Evaluation Logic</h2>
 * <ul>
 *   <li>If the flag is null or disabled, returns {@code null} (user not in rollout).</li>
 *   <li>If {@code hash >= rolloutPercentage/100.0}, user is outside the rollout → {@code null}.</li>
 *   <li>For boolean flags (no variants), returns {@code "on"} for users in rollout.</li>
 *   <li>For multi-variant flags, picks a variant proportionally using the same hash
 *       re-scaled within the rollout band.</li>
 * </ul>
 */
public class FeatureFlagEvaluator {

    /**
     * Maximum unsigned 32-bit value + 1 (2^32). Used as the divisor to normalize the hash.
     * Matches Python: {@code MAX_HASH_VALUE + 1 == 0x100000000}.
     */
    static final long HASH_DIVISOR = 0x100000000L;

    /**
     * Computes a normalized hash value in [0.0, 1.0) for the given user ID and salt.
     *
     * <p>The algorithm matches {@code ConsistentHasher._hash()} and
     * {@code ConsistentHasher.get_normalized_hash()} from the Python Lambda:
     * <pre>
     *     combined = f"{user_id}:{salt}".encode('utf-8')
     *     hash_bytes = hashlib.md5(combined).digest()[:4]
     *     hash_value = struct.unpack('&lt;I', hash_bytes)[0]   # little-endian unsigned int
     *     return hash_value / (MAX_HASH_VALUE + 1)          # normalize
     * </pre>
     *
     * @param userId  the user's unique identifier
     * @param salt    the salt string (flag key, experiment key, or suffixed key)
     * @return normalized hash in [0.0, 1.0)
     * @throws RuntimeException if MD5 is unexpectedly unavailable (never in practice)
     */
    public double computeHash(String userId, String salt) {
        try {
            String input = userId + ":" + salt;
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] bytes = md.digest(input.getBytes(StandardCharsets.UTF_8));

            // Little-endian unsigned 32-bit integer from first 4 bytes.
            // Python: struct.unpack('<I', hash_bytes)[0]
            // '<I' = little-endian unsigned int:
            //   byte[0] is least significant, byte[3] is most significant.
            long hash = ((bytes[0] & 0xFFL))
                      | ((bytes[1] & 0xFFL) << 8)
                      | ((bytes[2] & 0xFFL) << 16)
                      | ((bytes[3] & 0xFFL) << 24);

            // Divide by 2^32 to normalize to [0.0, 1.0)
            return hash / (double) HASH_DIVISOR;
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("MD5 MessageDigest not available on this JVM", e);
        }
    }

    /**
     * Evaluates whether a user should receive a feature flag, and which variant.
     *
     * <p>Returns {@code null} if the flag is disabled or the user is outside the rollout.
     * Returns the variant name if the user is in rollout (e.g., {@code "on"}, {@code "control"},
     * {@code "treatment"}).
     *
     * @param user the user to evaluate
     * @param flag the feature flag configuration
     * @return variant name, or {@code null} if the user is not in rollout
     */
    public String evaluate(User user, FeatureFlag flag) {
        if (flag == null || !flag.isEnabled()) {
            return null;
        }

        double hash = computeHash(user.getUserId(), flag.getKey());
        double rolloutFraction = flag.getRolloutPercentage() / 100.0;

        // User is outside the rollout band
        if (hash >= rolloutFraction) {
            return null;
        }

        List<FeatureFlag.Variant> variants = flag.getVariants();

        // Boolean flag: no variants defined → return "on"
        if (variants == null || variants.isEmpty()) {
            return "on";
        }

        // Multi-variant: re-scale hash within the rollout band [0, rolloutFraction)
        // to [0.0, 1.0) and pick a variant proportionally by weight.
        double variantHash = (rolloutFraction > 0.0) ? (hash / rolloutFraction) : 0.0;

        double cumulative = 0.0;
        for (FeatureFlag.Variant variant : variants) {
            cumulative += variant.getWeight();
            if (variantHash < cumulative) {
                return variant.getName();
            }
        }

        // Fallback to last variant in case of floating-point rounding
        return variants.get(variants.size() - 1).getName();
    }
}

File: sdk/ios/Sources/ExperimentationSDK/FeatureFlagEvaluator.swift
import Foundation
import CommonCrypto

/// Local feature flag evaluator using MD5-based consistent hashing.
///
/// ## Hash Algorithm
/// The algorithm is byte-for-byte compatible with all other SDK implementations
/// (Java, Python, JavaScript) and the backend Lambda consistent_hash.py:
///
/// 1. Compute MD5 of `"{userId}:{flagKey}"` encoded as UTF-8.
/// 2. Read the first 4 bytes as a little-endian unsigned 32-bit integer.
/// 3. Divide by `0x100000000` (2^32) to normalize to `[0.0, 1.0)`.
///
/// ## Evaluation Logic
/// - If the flag is disabled globally, returns `EvalResult(enabled: false, reason: "flag_disabled")`.
/// - If `hash >= rolloutPercentage / 100.0`, the user is outside the rollout band → `enabled: false`.
/// - For boolean flags (no variants), returns `enabled: true` with `reason: "in_rollout"`.
/// - For multi-variant flags, the hash is re-scaled within the rollout band to pick a variant
///   proportionally by weight.
public struct FeatureFlagEvaluator {

    /// Divisor used to normalize the 32-bit hash value into [0.0, 1.0).
    /// Equals 2^32 = 4294967296, matching Python's `MAX_HASH_VALUE + 1`.
    static let hashDivisor: Double = 0x100000000  // 4294967296.0

    // MARK: - Public Interface

    /// Computes a deterministic, normalized hash value in `[0.0, 1.0)` for the
    /// given user ID and flag key combination.
    ///
    /// The hash is cross-SDK compatible: given the same inputs, Java, Python, JS,
    /// and Swift implementations all produce the same output.
    ///
    /// - Parameters:
    ///   - userId: The user's unique identifier.
    ///   - flagKey: The feature flag's unique key (used as salt).
    /// - Returns: A value in `[0.0, 1.0)`.
    public static func hashUser(_ userId: String, flagKey: String) -> Double {
        let input = "\(userId):\(flagKey)"
        guard let data = input.data(using: .utf8) else { return 0.0 }

        var digest = [UInt8](repeating: 0, count: Int(CC_MD5_DIGEST_LENGTH))
        data.withUnsafeBytes { ptr in
            _ = CC_MD5(ptr.baseAddress, CC_LONG(data.count), &digest)
        }

        // Read first 4 bytes as little-endian unsigned 32-bit integer.
        // This matches Java:  ((bytes[0] & 0xFF)) | ((bytes[1] & 0xFF) << 8) | ...
        // And Python: struct.unpack('<I', hash_bytes[:4])[0]
        let v = UInt32(digest[0])
            | UInt32(digest[1]) << 8
            | UInt32(digest[2]) << 16
            | UInt32(digest[3]) << 24

        // Divide by 2^32 (not 2^32 - 1) to normalize to [0.0, 1.0).
        // This matches Java's HASH_DIVISOR = 0x100000000L and Python's MAX_HASH_VALUE + 1.
        return Double(v) / hashDivisor
    }

    /// Evaluates whether a feature flag is enabled for a given user and, if applicable,
    /// which variant they are assigned to.
    ///
    /// - Parameters:
    ///   - flag: The feature flag configuration from the API.
    ///   - user: The user to evaluate.
    /// - Returns: An `EvalResult` describing the evaluation outcome.
    public static func evaluate(_ flag: FeatureFlag, user: User) -> EvalResult {
        // Fast path: flag globally disabled.
        guard flag.enabled else {
            return EvalResult(enabled: false, variantKey: nil, value: nil, reason: "flag_disabled")
        }

        let hash = hashUser(user.id, flagKey: flag.key)
        let rolloutFraction = flag.rolloutPercentage / 100.0

        // User's hash falls outside the rollout band.
        guard hash < rolloutFraction else {
            return EvalResult(enabled: false, variantKey: nil, value: nil, reason: "out_of_rollout")
        }

        // Multi-variant flag: assign a variant proportionally.
        if let variants = flag.variants, !variants.isEmpty {
            let variantKey = assignVariant(variants, hash: hash, rolloutFraction: rolloutFraction)
            return EvalResult(
                enabled: true,
                variantKey: variantKey,
                value: nil,
                reason: "variant_assigned"
            )
        }

        // Boolean flag: user is simply in rollout.
        return EvalResult(enabled: true, variantKey: nil, value: nil, reason: "in_rollout")
    }

    // MARK: - Private Helpers

    /// Picks a variant for a user who is within the rollout band.
    ///
    /// The user's hash is re-scaled from the rollout band `[0, rolloutFraction)` to
    /// `[0.0, 1.0)` and compared against cumulative variant weights.
    ///
    /// - Parameters:
    ///   - variants: The list of variants with relative weights.
    ///   - hash: The user's normalized hash value.
    ///   - rolloutFraction: The rollout fraction (rolloutPercentage / 100).
    /// - Returns: The key of the selected variant.
    private static func assignVariant(
        _ variants: [Variant],
        hash: Double,
        rolloutFraction: Double
    ) -> String {
        // Normalize the hash to [0.0, 1.0) within the rollout band.
        let normalized = rolloutFraction > 0.0 ? hash / rolloutFraction : 0.0

        let totalWeight = variants.reduce(0.0) { $0 + $1.weight }
        var cumulative = 0.0

        for variant in variants {
            cumulative += variant.weight / totalWeight
            if normalized < cumulative {
                return variant.key
            }
        }

        // Fallback to last variant for floating-point edge cases.
        return variants.last?.key ?? ""
    }
}
